src/ripple/overlay/impl/Message.cpp

```diff
@@ -36,23 +40,86 @@ Message::Message (::google::protobuf::Message const& message, int type)
 
     assert (messageBytes != 0);
 
-    /** Number of bytes in a message header. */
-    std::size_t constexpr headerBytes = 6;
-
     mBuffer.resize (headerBytes + messageBytes);
 
-    auto ptr = mBuffer.data();
+    setHeader(mBuffer.data(), messageBytes, type, Compressed::Off);
 
-    *ptr++ = static_cast<std::uint8_t>((messageBytes >> 24) & 0xFF);
-    *ptr++ = static_cast<std::uint8_t>((messageBytes >> 16) & 0xFF);
-    *ptr++ = static_cast<std::uint8_t>((messageBytes >> 8) & 0xFF);
-    *ptr++ = static_cast<std::uint8_t>(messageBytes & 0xFF);
+    if (messageBytes != 0)
+        message.SerializeToArray(mBuffer.data() + headerBytes, messageBytes);
+}
 
-    *ptr++ = static_cast<std::uint8_t>((type >> 8) & 0xFF);
-    *ptr++ = static_cast<std::uint8_t> (type & 0xFF);
+void
+Message::compress()
+{
+    auto const messageBytes = mBuffer.size () - headerBytes;
 
-    if (messageBytes != 0)
-        message.SerializeToArray(ptr, messageBytes);
+    auto type = getType(mBuffer.data());
+
+    bool const compressible = [&]{
+        if (messageBytes <= 70)
+            return false;
+        switch(type)
+        {
+            case protocol::mtMANIFESTS:
+            case protocol::mtENDPOINTS:
+            case protocol::mtTRANSACTION:
+            case protocol::mtGET_LEDGER:
+            case protocol::mtLEDGER_DATA:
+            case protocol::mtGET_OBJECTS:
+            case protocol::mtVALIDATORLIST:
+                return true;
+            default:
+                return false;
+        }
+    }();
+
+    if (compressible)
+    {
+        auto *payload = static_cast<void const*>(mBuffer.data() + headerBytes);
+
+        auto compressedSize = ripple::compression::compress(
+                payload,
+                messageBytes,
+                [&](std::size_t in_size) { // size of required compressed buffer
+                    mBufferCompressed.resize(in_size + headerBytes);
+                    return (mBufferCompressed.data() + headerBytes);
+                });
+
+        double ratio = 1.0 - (double)compressedSize / (double)messageBytes;
+
+        // TODO should we have min acceptable compression ratio?
+        if (ratio > 0.0)
+        {
+            mBufferCompressed.resize(headerBytes + compressedSize);
+            setHeader(mBufferCompressed.data(), compressedSize, type, Compressed::On);
+        }
+        else
+            mBufferCompressed.resize(0);
+    }
+
+    mCompressedRequested = true;
+}
+
+/** Set payload header
+ * 47       Set to 1, indicates the message is compressed
+ * 46-44    Compression algorithm, value 1-7. Set to 1 to indicate LZ4 compression
+ * 43-42    Set to 0
+ * 41-16    Payload size
+ * 15-0	    Message Type
+*/
+void
+Message::setHeader(std::uint8_t *in, uint32_t messageBytes, int type,
+                   Compressed compressed, std::uint8_t comprAlgorithm)
+{
+    uint8_t compression = (compressed == Compressed::On?0xF0:0) & (0x80 | (comprAlgorithm << 4));
```

Comment:
The main motivation for using `varint` is to save a couple of bytes since the vast majority of the messages are small. We could move the original size into the message's header but then it has to be fixed at 4 bytes.